extension/extensionmiddleware/server.go

```diff
@@ -0,0 +1,29 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package extensionmiddleware // import "go.opentelemetry.io/collector/extension/extensionmiddleware"
+
+import (
+	"net/http"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/stats"
+)
+
+// HTTPServer defines the interface for HTTP server middleware extensions.
+type HTTPServer interface {
+	// Handler wraps the provided base http.Handler.
+	ServerHandler(base http.Handler) (http.Handler, error)
+}
+
+// GRPCServer defines the interface for gRPC server middleware extensions.
+type GRPCServer interface {
+	// ServerUnaryInterceptor returns a gRPC unary server interceptor.
+	ServerUnaryInterceptor() (grpc.UnaryServerInterceptor, error)
+
+	// ServerStreamInterceptor returns a gRPC stream server interceptor.
+	ServerStreamInterceptor() (grpc.StreamServerInterceptor, error)
+
+	// ServerStatsHandler returns a gRPC stats handler for server-side operations.
+	ServerStatsHandler() (stats.Handler, error)
```

Comment:
Same question as for client, not sure if this should be here or a separate interface. Idea is that in general you either have implementations for the 2 interceptors or for the stats handler.

Reply:
As noted above, I've added to extensionlimiter a new Provider interface, which returns RateLimiter or ResourceLimiters on request. This means that the stats handler and/or interceptor are only created as needed.

Reply:
Fixed. Now there are four interfaces (HTTP, GRPC) x (Client, Server) and they return the type of the corresponding option or interceptor. E.g., returns []grpc.ServerOption for gRPC servers.